### PR TITLE
Herbert polish model

### PR DESCRIPTION
**Pull request by rmroczkowski**
The HerBERT model is a transformer model pretrained using masked language modeling (MLM) and Sentence Structural (SSO) objectives for the Polish language. 

It was added to the library in PyTorch with the following checkpoints:

- `allegro/herbert-base-cased`
- `allegro/herbert-large-cased`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dimiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), 
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests? 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.
@patrickvonplaten
@julien-c
@LysandreJik

### CODE REVIEW

**Review by LysandreJik — COMMENTED** [review]
Hello. thanks a lot for your contribution! It seems that your model is entirely based on the BERT model, with no modification to its code. It seems the major difference is the tokenizer, which is based on the XLM tokenizer with a few modified aspects.

In that case, could you leverage the following PR: https://github.com/huggingface/transformers/pull/6995

Instead of redefining all HerBert models which are basically inheriting from BERT, you would simply load the model checkpoint in the BERT architecture. However, there would still be a `HerbertTokenizer` as this requires additional code.

You can take inspiration from the [Phobert model](https://huggingface.co/vinai/phobert-base#). See how its [configuration file](https://s3.amazonaws.com/models.huggingface.co/bert/vinai/phobert-base/config.json) mentions that it is based on a RoBERTa implementation, but still leverages a `PhobertTokenizer `. You can see its implementation details [here](https://github.com/huggingface/transformers/pull/6129).

**Comment by rmroczkowski** [comment]
Great point! I've already simplified the code and left on;y the tokenizer. Ideally, `XLMTokenizer` should have a converter to the appropriate `Fast` class. On second thought, I can see the problem would be to program the Moses pretokenization in the `tokenizers` library.

**Review by LysandreJik — APPROVED** [review]
The updated code looks great! Thanks for implementing it. I'll take a look at how to add it to the docs.

File: model_cards/allegro/herbert-base-cased/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+---
+language: pl
+tags:
+- pytorch
+- herbert
+- masked-lm
```
Comment (julien-c) [inline model_cards/allegro/herbert-base-cased/README.md:6]:
```suggestion
- herbert
```

You don't need those two, they'll be added automatically

File: model_cards/allegro/herbert-large-cased/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+---
+language: pl
+tags:
+- pytorch
+- herbert
+- masked-lm
```
Comment (julien-c) [inline model_cards/allegro/herbert-large-cased/README.md:6]:
```suggestion
- herbert
```

File: src/transformers/tokenization_herbert.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    "merges_file": {
+        "allegro/herbert-base-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/allegro/herbert-base-cased/merges.txt",
+        "allegro/herbert-large-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/allegro/herbert-large-cased/merges.txt",
+    },
+}
+
+PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {
+    "allegro/herbert-base-cased": 512,
+    "allegro/herbert-large-cased": 512,
+}
```
Comment (julien-c) [inline src/transformers/tokenization_herbert.py:45]: You shouldn't need to hardcode those anymore, but I understand that some code in the repo still uses that at the moment.

Comment (rmroczkowski) [reply src/transformers/tokenization_herbert.py]: Deprecated configuration removed in next commit

**Review by LysandreJik — APPROVED** [review]
Thanks for updating!

**Comment by datquocnguyen** [comment]
Hi @rmroczkowski 

As far as I know, you employed fastBPE to train a tokenizer: https://github.com/huggingface/transformers/tree/b23d3a5ad4aa08decd10671f85be5950767dd052/model_cards/allegro/herbert-klej-cased-tokenizer-v1

I also employed fastBPE for the Vietnamese BERT-based tokenizer (i.e. PhoBERTTokenizer https://github.com/huggingface/transformers/pull/6129 ), but I am still struggling to implement a fast tokenizer based on fastBPE, e.g. handling the suffix "@@" of subword tokens. In particular, given https://huggingface.co/vinai/phobert-base/tree/main I can convert "bpe.codes" into a "merge.txt"-style file, but I am not sure about how to convert our "vocab.txt" into your "vocab.json"-style file.

How can you convert your fastBPE's code and vocab outputs into HuggingFace's tokenizers? So that you can call the tokenizer with the use_fast=True option.  

cc: @LysandreJik  is there any idea for implementing a fast version of a fastBPE-based slow one?

Thank you both.
